File: rust-martian-ddd/martian-robot-ddd/src/application/mission.rs
use crate::application::{in_memory_scents::InMemoryScents, parsing};
use crate::domain::navigator::execute;
use crate::domain::{grid::Grid, orientation::Orientation, robot::Robot};
use std::io::{BufRead, Write};

fn read_non_empty<R: BufRead>(r: &mut R, buf: &mut String) -> Option<String> {
    buf.clear();
    loop {
        let n = r.read_line(buf).ok()?;
        if n == 0 {
            return None;
        }
        let s = buf.trim();
        if !s.is_empty() {
            return Some(s.to_string());
        }
        buf.clear();
    }
}

pub fn run<R: BufRead, W: Write>(reader: &mut R, mut writer: W) -> std::io::Result<()> {
    let mut buf = String::new();
    let header = match read_non_empty(reader, &mut buf) {
        Some(s) => s,
        None => return Ok(()),
    };
    let (mx, my) = parsing::parse_grid(&header).expect("invalid grid header");
    let grid = Grid::new(mx, my);
    let mut scents = InMemoryScents::default();

    while let Some(meta) = read_non_empty(reader, &mut buf) {
        let (pos, o) = parsing::parse_robot(&meta).expect("invalid robot meta");
        let instr = read_non_empty(reader, &mut buf).unwrap_or_default();
        let instr = parsing::normalize_instr(&instr);

        let mut robot = Robot::new(pos, o);
        execute(&mut scents, &grid, &mut robot, &instr);

        writeln!(
            writer,
            "{} {} {}{}",
            robot.position.x,
            robot.position.y,
            orientation_to_str(robot.orientation),
            if robot.lost { " LOST" } else { "" }
        )?;
    }
    // trailing blank line to match goldens
    writeln!(writer)?;
    Ok(())
}

fn orientation_to_str(o: Orientation) -> &'static str {
    match o {
        Orientation::N => "N",
        Orientation::E => "E",
        Orientation::S => "S",
        Orientation::W => "W",
    }
}

File: langs/go-ddd/internal/app/mission.go
package app

import (
	"bufio"
	"fmt"
	"io"
	"martian-robot-ddd/internal/domain"
	"strconv"
	"strings"
)

func Run(r *bufio.Reader, w io.Writer) error {
	header, err := readNonEmpty(r)
	if err == io.EOF {
		return nil
	}
	if err != nil {
		return err
	}
	mx, my, err := parseGrid(header)
	if err != nil {
		return err
	}
	grid := domain.Grid{MaxX: mx, MaxY: my}
	scents := NewInMemoryScents()

	for {
		meta, err := readNonEmpty(r)
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
		parts := strings.Fields(meta)
		x := atoi(parts[0])
		y := atoi(parts[1])
		o := domain.Orientation(parts[2])

		instr, err := readNonEmpty(r)
		if err != nil && err != io.EOF {
			return err
		}
		instr = filterInstr(instr)

		robot := &domain.Robot{Position: domain.Position{X: x, Y: y}, Orientation: o}
		domain.Execute(scents, grid, robot, instr)

		if robot.Lost {
			fmt.Fprintf(w, "%d %d %s LOST\n", robot.Position.X, robot.Position.Y, robot.Orientation)
		} else {
			fmt.Fprintf(w, "%d %d %s\n", robot.Position.X, robot.Position.Y, robot.Orientation)
		}
		if err == io.EOF {
			break
		}
	}
	// trailing newline to match goldens
	fmt.Fprintln(w)
	return nil
}

func atoi(s string) int { n, _ := strconv.Atoi(s); return n }

func filterInstr(s string) string {
	var b strings.Builder
	for _, c := range s {
		if c == 'L' || c == 'R' || c == 'F' {
			b.WriteRune(c)
		}
	}
	return b.String()
}
